<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Lab 7</title>
</head>

<body>

    <div>
        <h1>Lab Assignment 7</h1>
        <hr>
        <h3>Question 1: Fix using wrapper, bind, call, apply.</h3>
        <div>
            <div>
                <pre>
                    function askPassword(ok, fail) {
                        let password = prompt("Password", '');
                        if (password == "rockstar") ok();
                        else fail();
                    }
                    
                    let user = {
                        name: 'John',
                        loginOk() {
                            alert(`${this.name} logged in`);
                        },
                        loginFailed() {
                            alert(`${this.name} failed login`)
                        },
                    };
                    askPassword(user.loginOk, user.loginFail);
                        
                </pre>
                <h3>Answer:</h3>
                <pre>
                    askPassword(user.loginOk, user.loginFailed); --- The problem occurs as the context of the object user is not found in the LE so here name is undefined 
                    
                    wrapper solution(Wrapper Solution here the context of the object is found!)
                    askPassword(() => user.loginOk(), () => user.loginFailed());
                    
                    bind solution(Bind Solution here the context of the object is bind with the helper bind function)
                    askPassword(user.loginOk.bind(user), user.loginFailed.bind(user));
                    
                    apply solution(apply solution here the context of the object is wrapped with apply function)
                    askPassword(() => user.loginOk.apply(user), () => user.loginFailed.apply(user));
                    
                    call context solution(call solution here the context of the object is wrapped with call function )
                    askPassword(() => user.loginOk.call(user), () => user.loginFailed.call(user));  
                            
                </pre>
            </div>
        </div>
    </div> 
        <hr>



        <div>
            <hr>
            <h3>Question 2: Fix the code below using bind</h3>
            <div>
                <div>
                    <pre>
                        let group = {
                            title: "Our Group",
                            students: ["John", "Pete", "Alice"],
                            showList: function() {
                            this.students.forEach(function(student) {
                            console.log(this.title + ": " + student
                                        );
                                    });
                                }
                            };
                            group.showList();
                            
                        </pre>
                    <h3>Answer:</h3>
                    <pre>
                        let group = {
                            title: "Our Group",
                            students: ["John", "Pete", "Alice"],
                            showList: function() {
                            this.students.forEach(function(student){
                            console.log(this.title + ": " + student);
                                    }.bind(this));
                                }
                            };
                            group.showList();
                                
                            </pre>
                </div>
            </div>
        </div>
    </body>

</html>